src/storage/UpdateEdgeProcessor.cpp

```diff
@@ -0,0 +1,382 @@
+    auto schema = std::const_pointer_cast<meta::SchemaProviderIf>(constSchema);
+    std::unique_ptr<RowUpdater> updater;
+    // TODO(zhangguoqing) When the edge is to be inserted, it can't instantiate
+    // the class RowUpdater() with edgeReader_ parameter, otherwise the updater
+    // doing encoded will report the error: Failed to encode updated data.
+    if (insertEdge_) {
+        updater = std::unique_ptr<RowUpdater>(new RowUpdater(schema));
+        updateRespData.set_upsert(true);
+    } else {
+        updater = std::unique_ptr<RowUpdater>(new RowUpdater(
+                    std::move(edgeReader_), schema));
+    }
+    for (auto& item : updateItems_) {
+        auto edgeName = item.get_name();
+        auto field = item.get_field();
+        StatusOr<std::unique_ptr<Expression>> exp = Expression::decode(item.get_value());
+        auto vexp = std::move(exp).value();
+        vexp->setContext(expCtx_.get());
+        auto expValue = vexp->eval();
+        // update the props and the record
+        int32_t retIndex = -1;
+        if (returnColumnsNum_ > 0) {
+            for (auto& prop : this->edgeContext_.props_) {
+                if (prop.prop_.name == field && prop.returned_) {
+                    retIndex = prop.retIndex_;
+                    break;
+                }
+            }
+        }
+        switch (expValue.which()) {
+            case VAR_INT64: {
+                auto v = boost::get<int64_t>(expValue);
+                updater->setInt(field, v);
+                if (retIndex != -1) record[retIndex] = v;
+                break;
+            }
+            case VAR_DOUBLE: {
+                auto v = boost::get<double>(expValue);
+                updater->setDouble(field, v);
+                if (retIndex != -1) record[retIndex] = v;
+                break;
+            }
+            case VAR_BOOL: {
+                auto v = boost::get<bool>(expValue);
+                updater->setBool(field, v);
+                if (retIndex != -1) record[retIndex] = v;
+                break;
+            }
+            case VAR_STR: {
+                auto v = boost::get<std::string>(expValue);
+                updater->setString(field, v);
+                if (retIndex != -1) record[retIndex] = v;
+                break;
+             }
+            default:
+                LOG(FATAL) << "Unknown VariantType: " << expValue.which();
+        }
+    }
+
+    // the updated data be written to kvstore
+    std::vector<kvstore::KV> data;
+    data.emplace_back(key_, std::move(updater->encode()));
+    {
+        folly::Baton<true, std::atomic> baton;
+        this->kvstore_->asyncMultiPut(spaceId_, partId, std::move(data),
+                [&baton, &ret, partId, this] (kvstore::ResultCode code) {
+            VLOG(3) << "spaceId: " << this->spaceId_ << ", partId: "
+                    << partId << ", code:" << static_cast<int32_t>(code);
+            ret = code;
+            baton.post();
+        });
+        baton.wait();
```

Comment:
maybe here can not wait directly, because asyncMultiPut will run raft, and it will take a long time, the wait will let a worker thread sleep, it will cut down the throughput of system

Reply:
If disable the baton.wait(), the test case will failure when checking the kvstore, because the updated data has not written to the kvstore.

Reply:
here can get a future and add a lambda to worker like function "RaftPart::replicateLogs":
get_future(xxx).then(worker, []{check the kvstore});